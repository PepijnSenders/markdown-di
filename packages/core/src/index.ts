import matter from 'gray-matter';
import type { ProcessOptions, ProcessResult, ProcessingContext, ValidationError } from './types';
import { FrontmatterValidator } from './validator';
import { ReferenceValidator, ReferenceExistenceValidator } from './validator';
import { DependencyResolver, CircularDependencyDetector } from './resolver';
import { ContentProcessor, FrontmatterProcessor } from './processor';
import { SchemaValidator, parseYamlSchema, validateInlineSchema } from './schema';
import { z } from 'zod';
import * as yaml from 'yaml';

/**
 * Main class for markdown dependency injection
 */
export class MarkdownDI {
  private frontmatterValidator = new FrontmatterValidator();
  private referenceValidator = new ReferenceValidator();
  private referenceExistenceValidator = new ReferenceExistenceValidator();
  private frontmatterProcessor = new FrontmatterProcessor();
  private schemaRegistry = new Map<string, z.ZodSchema<any>>();

  /**
   * Register a schema by name
   */
  registerSchema(name: string, schema: z.ZodSchema<any>): void {
    this.schemaRegistry.set(name, schema);
  }

  /**
   * Register schemas from configuration object
   */
  registerSchemas(schemas: Record<string, z.ZodSchema<any>>): void {
    Object.entries(schemas).forEach(([name, schema]) => {
      this.registerSchema(name, schema);
    });
  }

  /**
   * Get a registered schema by name
   */
  getSchema(name: string): z.ZodSchema<any> | undefined {
    return this.schemaRegistry.get(name);
  }

  /**
   * Process markdown content with dependency injection
   */
  async process(options: ProcessOptions): Promise<ProcessResult> {
    const context: ProcessingContext = {
      baseDir: options.baseDir,
      mode: options.mode || 'build',
      visitedFiles: new Set(),
      currentFile: options.currentFile
    };

    // Register any schemas from options
    if (options.schemas) {
      this.registerSchemas(options.schemas);
    }

    // Extract and validate frontmatter
    let { frontmatter, body, errors: frontmatterErrors } = this.frontmatterProcessor.extract(options.content);

    // Combine all errors
    const allErrors: ProcessResult['errors'] = [...frontmatterErrors];

    // Validate frontmatter structure
    const frontmatterValidationErrors = this.frontmatterValidator.validate(frontmatter);
    allErrors.push(...frontmatterValidationErrors);

    // Handle schema prop in frontmatter
    let validationSchema: z.ZodSchema<any> | null = null;

    if (frontmatter.schema) {
      if (typeof frontmatter.schema === 'string') {
        // Reference to registered schema
        const registeredSchema = this.getSchema(frontmatter.schema);
        if (registeredSchema) {
          validationSchema = registeredSchema;
        } else {
          allErrors.push({
            type: 'schema',
            message: `Referenced schema '${frontmatter.schema}' is not registered`,
            location: 'schema'
          });
        }
      } else if (typeof frontmatter.schema === 'object') {
        // Inline schema definition
        try {
          validationSchema = parseYamlSchema(yaml.stringify(frontmatter.schema));
        } catch (error) {
          allErrors.push({
            type: 'schema',
            message: `Invalid inline schema: ${error instanceof Error ? error.message : String(error)}`,
            location: 'schema'
          });
        }
      }
    }

    // Validate against schema if provided (from frontmatter or options)
    if (validationSchema) {
      const schemaValidator = new SchemaValidator({ schema: validationSchema });
      const schemaResult = schemaValidator.validate(frontmatter);

      if (!schemaResult.valid) {
        allErrors.push(...schemaResult.errors);
      } else {
        // Update frontmatter with validated/transformed data
        frontmatter = schemaResult.data;
      }
    } else if (options.schema) {
      // Fallback to CLI-provided schema
      const schemaValidator = new SchemaValidator(options.schema);
      const schemaResult = schemaValidator.validate(frontmatter);

      if (!schemaResult.valid) {
        allErrors.push(...schemaResult.errors);
      } else {
        // Update frontmatter with validated/transformed data
        frontmatter = schemaResult.data;
      }
    }

    // Validate reference syntax in content
    const syntaxValidationErrors = this.referenceValidator.validate(body);
    allErrors.push(...syntaxValidationErrors);

    // Validate that references exist in frontmatter
    const existenceValidationErrors = this.referenceExistenceValidator.validate(body, frontmatter);
    allErrors.push(...existenceValidationErrors);

    // Validate that injected variables are defined in frontmatter and schema
    const injectionValidationErrors = this.validateInjectionVariables(body, frontmatter, validationSchema || undefined);
    allErrors.push(...injectionValidationErrors);

    // Initialize resolver and detector
    const resolver = new DependencyResolver(context);
    const circularDetector = new CircularDependencyDetector();

    // Resolve dependencies and check for circular references
    const { dependencies, errors: resolutionErrors } = resolver.resolve(frontmatter);
    allErrors.push(...resolutionErrors);

    // Detect circular dependencies
    if (context.currentFile) {
      const circularErrors = circularDetector.detect(context.currentFile, dependencies);
      allErrors.push(...circularErrors);
    }

    // Process content if no errors or in build mode
    let processedContent = body;
    if (allErrors.length === 0 || context.mode === 'build') {
      const processor = new ContentProcessor(resolver, circularDetector);
      const { processedContent: finalContent, errors: processingErrors } = await processor.process(
        body,
        frontmatter,
        context
      );
      processedContent = finalContent;
      allErrors.push(...processingErrors);
    }

    // Reassemble the document with processed frontmatter
    let finalContent;
    try {
      finalContent = this.reassembleDocument(frontmatter, processedContent);
    } catch (error) {
      allErrors.push({
        type: 'frontmatter',
        message: `Failed to reassemble document: ${error}`,
        location: 'document'
      });
      finalContent = options.content; // Return original content if reassembly fails
    }

    return {
      content: finalContent,
      frontmatter,
      errors: allErrors,
      dependencies
    };
  }

  /**
   * Validate that injected variables are defined in frontmatter
   */
  private validateInjectionVariables(body: string, frontmatter: any, validationSchema?: z.ZodSchema<any>): ValidationError[] {
    const errors: ValidationError[] = [];

    // Extract all references from content
    const referenceRegex = /\{\{([^}]+)\}\}/g;
    const matches: string[] = [];
    let match;

    while ((match = referenceRegex.exec(body)) !== null) {
      matches.push(match[1].trim());
    }

    // Get all available variables from frontmatter
    const availableVariables = new Set<string>();

    // Add standard frontmatter fields
    Object.keys(frontmatter).forEach(key => {
      availableVariables.add(key);
    });

    // Add blueprint variables (blueprints.group.key)
    if (frontmatter.blueprints) {
      Object.entries(frontmatter.blueprints).forEach(([group, items]) => {
        Object.entries(items as Record<string, string>).forEach(([key]) => {
          availableVariables.add(`${group}.${key}`);
        });
      });
    }

    // Add reference variables (references.group)
    if (frontmatter.references) {
      Object.entries(frontmatter.references).forEach(([group]) => {
        availableVariables.add(group);
      });
    }

    // Check each reference
    matches.forEach(reference => {
      const parts = reference.split('.');
      const root = parts[0];

      if (!availableVariables.has(reference) && !availableVariables.has(root)) {
        errors.push({
          type: 'injection',
          message: `Injected variable '${reference}' is not defined in frontmatter`,
          location: 'content'
        });
      }
    });

    // If there's a validation schema, ensure all required fields are present
    if (validationSchema) {
      try {
        const result = validationSchema.safeParse(frontmatter);
        if (!result.success) {
          result.error.errors.forEach((issue: any) => {
            errors.push({
              type: 'schema',
              message: `Schema validation failed: ${issue.message}`,
              location: issue.path.join('.')
            });
          });
        }
      } catch (error) {
        errors.push({
          type: 'schema',
          message: `Schema validation error: ${error instanceof Error ? error.message : String(error)}`,
          location: 'schema'
        });
      }
    }

    return errors;
  }

  /**
   * Validate markdown content without processing
   */
  async validate(options: Omit<ProcessOptions, 'mode'>): Promise<ProcessResult> {
    return this.process({ ...options, mode: 'validate' });
  }

  /**
   * Reassemble document with frontmatter and processed body
   */
  private reassembleDocument(frontmatter: any, body: string): string {
    // Convert frontmatter back to YAML
    const frontmatterYaml = this.stringifyFrontmatter(frontmatter);

    // Reassemble the document
    return `---\n${frontmatterYaml}\n---\n${body}`;
  }

  /**
   * Convert frontmatter object to YAML string
   */
  private stringifyFrontmatter(frontmatter: any): string {
    const lines: string[] = [];

    // Sort keys for consistent output
    const sortedKeys = Object.keys(frontmatter).sort((a, b) => {
      // Required fields first
      if (a === 'name') return -1;
      if (b === 'name') return 1;
      if (a === 'description') return -1;
      if (b === 'description') return 1;
      // Then blueprints and references
      if (a === 'blueprints') return -1;
      if (b === 'blueprints') return 1;
      if (a === 'references') return -1;
      if (b === 'references') return 1;
      return a.localeCompare(b);
    });

    for (const key of sortedKeys) {
      const value = frontmatter[key];

      if (value === undefined || value === null) continue;

      if (typeof value === 'object' && !Array.isArray(value)) {
        // Handle objects
        lines.push(`${key}:`);
        this.stringifyObject(value, lines, 1);
      } else if (Array.isArray(value)) {
        // Handle arrays
        lines.push(`${key}:`);
        for (const item of value) {
          if (typeof item === 'string') {
            lines.push(`  - ${item}`);
          } else {
            lines.push(`  - ${JSON.stringify(item)}`);
          }
        }
      } else {
        // Handle primitives
        const stringValue = typeof value === 'string' ? `"${value.replace(/"/g, '\\"')}"` : value;
        lines.push(`${key}: ${stringValue}`);
      }
    }

    return lines.join('\n');
  }

  /**
   * Recursively stringify nested objects
   */
  private stringifyObject(obj: any, lines: string[], indent: number): void {
    const spaces = '  '.repeat(indent);
    const sortedKeys = Object.keys(obj).sort();

    for (const key of sortedKeys) {
      const value = obj[key];

      if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
        lines.push(`${spaces}${key}:`);
        this.stringifyObject(value, lines, indent + 1);
      } else if (Array.isArray(value)) {
        lines.push(`${spaces}${key}:`);
        for (const item of value) {
          lines.push(`${spaces}  - ${item}`);
        }
      } else {
        const stringValue = typeof value === 'string' ? `"${value.replace(/"/g, '\\"')}"` : value;
        lines.push(`${spaces}${key}: ${stringValue}`);
      }
    }
  }
}

// Export types and classes
export * from './types';
export * from './validator';
export * from './resolver';
export * from './processor';

// Export the main class as default
/**
 * Schema utilities
 */
export class MarkdownDISchema {
  /**
   * Create a schema validator
   */
  static create(schema: z.ZodSchema<any>, options?: {
    extend?: boolean;
    strict?: boolean;
  }): SchemaValidator {
    return new SchemaValidator({ schema, ...options });
  }

  /**
   * Validate a single file with schema
   */
  static async validateFile(filePath: string, schema?: any, options?: {
    extend?: boolean;
    strict?: boolean;
  }) {
    const validator = new SchemaValidator({ schema, ...options });
    return validator.validateFile(filePath);
  }

  /**
   * Generate TypeScript type from schema
   */
  static generateType(schema: any, typeName = 'Frontmatter'): string {
    const validator = new SchemaValidator({ schema });
    return validator.generateType(schema, typeName);
  }
}

export default MarkdownDI;