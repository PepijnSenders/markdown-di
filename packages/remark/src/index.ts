import type { Plugin } from 'unified';
import type { Root } from 'mdast';
import { MarkdownDI } from '@markdown-di/core';

export interface RemarkMarkdownDIOptions {
  baseDir: string;
  /**
   * Enable heading level adjustments during transclusion
   */
  headingShift?: boolean;
  /**
   * Enable link rewriting for transcluded content
   */
  linkRewrite?: boolean;
  /**
   * Processing mode: 'validate' or 'build'
   */
  mode?: 'validate' | 'build';
}

/**
 * Remark plugin for markdown dependency injection with AST-based transformations
 */
export const remarkMarkdownDI: Plugin<[RemarkMarkdownDIOptions?], Root> = (options?: RemarkMarkdownDIOptions) => {
  const mdi = new MarkdownDI();

  return async (tree, file) => {
    const baseDir = options?.baseDir || process.cwd();

    const result = await mdi.process({
      content: file.value as string,
      baseDir,
      currentFile: file.path,
      mode: options?.mode
    });

    // Update the file with processed content
    file.value = result.content;

    // Store processing metadata
    file.data = {
      ...file.data,
      markdownDI: {
        frontmatter: result.frontmatter,
        dependencies: result.dependencies,
        errors: result.errors
      }
    };

    if (result.errors.length > 0) {
      const errorMessages = result.errors
        .map(error => `${error.type}: ${error.message} at ${error.location}`)
        .join('\n');
      file.messages.push({
        name: 'warning',
        message: `Markdown DI processing errors:\n${errorMessages}`
      } as any);
    }

    // Future: AST-level transformations for heading shifting and link rewriting
    if (options?.headingShift || options?.linkRewrite) {
      // TODO: Implement AST transformations
      // This would require parsing the transcluded content and manipulating the AST
    }
  };
};

export default remarkMarkdownDI;