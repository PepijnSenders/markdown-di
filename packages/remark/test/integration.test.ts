import { describe, test, expect, beforeAll, afterAll } from 'bun:test';
import { unified } from 'unified';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import remarkFrontmatter from 'remark-frontmatter';
import { remarkMarkdownDI } from '../src/index';
import { join } from 'path';
import { mkdirSync, writeFileSync, rmSync } from 'fs';

const TEST_DIR = join(__dirname, '../../test-fixtures-remark');

// Setup test fixtures
function setupTestFixtures() {
  // Create test directory
  mkdirSync(TEST_DIR, { recursive: true });
  mkdirSync(join(TEST_DIR, 'sections'), { recursive: true });
  mkdirSync(join(TEST_DIR, 'components'), { recursive: true });

  // Create section files
  writeFileSync(
    join(TEST_DIR, 'sections', 'header.md'),
    '# Header Section\n\nThis is a reusable header component.'
  );

  writeFileSync(
    join(TEST_DIR, 'sections', 'footer.md'),
    '---\n\nÂ© 2025 Test Company. All rights reserved.'
  );

  // Create component files
  writeFileSync(
    join(TEST_DIR, 'components', 'callout.md'),
    '> **Note:** This is an important callout message.'
  );

  writeFileSync(
    join(TEST_DIR, 'components', 'code-example.md'),
    '```javascript\nconst example = "Hello World";\nconsole.log(example);\n```'
  );
}

// Cleanup test fixtures
function cleanupTestFixtures() {
  rmSync(TEST_DIR, { recursive: true, force: true });
}

describe('remarkMarkdownDI - Integration Tests', () => {
  beforeAll(() => {
    setupTestFixtures();
  });

  afterAll(() => {
    cleanupTestFixtures();
  });

  describe('Basic Remark Processing', () => {
    test('processes markdown with dependency injection via remark', async () => {
      const input = `---
name: remark-test
description: Testing remark plugin

blueprints:
  sections:
    header: sections/header.md
---

{{sections.header}}

## Main Content

This is the main content.
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const result = await processor.process(input);

      expect(String(result)).toMatchSnapshot();
    });

    test('processes document with multiple transclusions', async () => {
      const input = `---
name: multi-transclusion
description: Multiple transclusions test

blueprints:
  sections:
    header: sections/header.md
    footer: sections/footer.md
  components:
    callout: components/callout.md
---

{{sections.header}}

## Content Section

Here's some content with a callout:

{{components.callout}}

More content here.

{{sections.footer}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const result = await processor.process(input);

      expect(String(result)).toMatchSnapshot();
    });

    test('handles code blocks in transclusions', async () => {
      const input = `---
name: code-transclusion
description: Testing code block transclusion

blueprints:
  components:
    example: components/code-example.md
---

# Code Examples

Here's a code example:

{{components.example}}

That's the example.
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const result = await processor.process(input);

      expect(String(result)).toMatchSnapshot();
    });
  });

  describe('Validation Mode', () => {
    test('validates without processing in validate mode', async () => {
      const input = `---
name: validation-test
description: Testing validation mode

blueprints:
  sections:
    header: sections/header.md
    nonexistent: sections/does-not-exist.md
---

{{sections.header}}
{{sections.nonexistent}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR, mode: 'validate' })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(String(vfile)).toMatchSnapshot();
      expect(vfile.messages.length).toBeGreaterThan(0);
      expect(vfile.data).toMatchSnapshot();
    });

    test('collects validation errors in file data', async () => {
      const input = `---
name: error-collection
description: Collecting validation errors
---

{{undefined.reference}}
{{another.missing}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR, mode: 'validate' })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(vfile.data).toMatchSnapshot();
      expect((vfile.data as any).markdownDI?.errors).toBeDefined();
      expect((vfile.data as any).markdownDI?.errors.length).toBeGreaterThan(0);
    });
  });

  describe('Build Mode', () => {
    test('processes and replaces references in build mode', async () => {
      const input = `---
name: build-mode-test
description: Testing build mode processing

blueprints:
  sections:
    header: sections/header.md
    footer: sections/footer.md
---

{{sections.header}}

# Main Section

Content goes here.

{{sections.footer}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR, mode: 'build' })
        .use(remarkStringify);

      const result = await processor.process(input);

      expect(String(result)).toMatchSnapshot();
    });
  });

  describe('Advanced Options', () => {
    test('respects headingShift option', async () => {
      const input = `---
name: heading-shift-test
description: Testing heading shift

blueprints:
  sections:
    header: sections/header.md
---

# Top Level

{{sections.header}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, {
          baseDir: TEST_DIR,
          headingShift: true
        })
        .use(remarkStringify);

      const result = await processor.process(input);

      // Note: This feature is planned but not yet implemented
      // The test documents the expected behavior
      expect(String(result)).toMatchSnapshot();
    });

    test('respects linkRewrite option', async () => {
      const input = `---
name: link-rewrite-test
description: Testing link rewriting

blueprints:
  sections:
    header: sections/header.md
---

# Document

{{sections.header}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, {
          baseDir: TEST_DIR,
          linkRewrite: true
        })
        .use(remarkStringify);

      const result = await processor.process(input);

      // Note: This feature is planned but not yet implemented
      // The test documents the expected behavior
      expect(String(result)).toMatchSnapshot();
    });
  });

  describe('File Metadata', () => {
    test('stores frontmatter in file data', async () => {
      const input = `---
name: metadata-test
description: Testing metadata storage
author: Test Author

blueprints:
  sections:
    header: sections/header.md
---

{{sections.header}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(vfile.data).toMatchSnapshot();
      expect((vfile.data as any).markdownDI?.frontmatter).toBeDefined();
      expect((vfile.data as any).markdownDI?.frontmatter.name).toBe('metadata-test');
      expect((vfile.data as any).markdownDI?.frontmatter.author).toBe('Test Author');
    });

    test('stores dependencies in file data', async () => {
      const input = `---
name: dependencies-test
description: Testing dependency tracking

blueprints:
  sections:
    header: sections/header.md
    footer: sections/footer.md
---

{{sections.header}}
{{sections.footer}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(vfile.data).toMatchSnapshot();
      expect((vfile.data as any).markdownDI?.dependencies).toBeDefined();
      expect((vfile.data as any).markdownDI?.dependencies.length).toBeGreaterThan(0);
    });

    test('stores errors in file data', async () => {
      const input = `---
name: errors-test
description: Testing error tracking

blueprints:
  sections:
    header: sections/header.md
---

{{sections.header}}
{{sections.undefined}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(vfile.data).toMatchSnapshot();
      expect((vfile.data as any).markdownDI?.errors).toBeDefined();
    });
  });

  describe('Complex Scenarios', () => {
    test('processes complex document with nested structures', async () => {
      const input = `---
name: complex-document
description: Complex document with various elements

blueprints:
  sections:
    header: sections/header.md
    footer: sections/footer.md
  components:
    callout: components/callout.md
    code: components/code-example.md
---

{{sections.header}}

## Introduction

This is a complex document demonstrating various features.

{{components.callout}}

## Code Examples

{{components.code}}

## Conclusion

Thank you for reading.

{{sections.footer}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const result = await processor.process(input);

      expect(String(result)).toMatchSnapshot();
      expect(result.data).toMatchSnapshot();
    });

    test('handles empty documents gracefully', async () => {
      const input = `---
name: empty-document
description: Empty document test
---
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const result = await processor.process(input);

      expect(String(result)).toMatchSnapshot();
    });

    test('preserves markdown formatting', async () => {
      const input = `---
name: formatting-preservation
description: Testing formatting preservation

blueprints:
  components:
    callout: components/callout.md
---

# Main Title

This is **bold** and this is *italic*.

{{components.callout}}

- List item 1
- List item 2
  - Nested item
- List item 3

1. Numbered list
2. Second item
   1. Nested numbered
3. Third item
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const result = await processor.process(input);

      expect(String(result)).toMatchSnapshot();
    });
  });

  describe('Error Handling', () => {
    test('handles missing frontmatter', async () => {
      const input = `# Document Without Frontmatter

{{sections.header}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(String(vfile)).toMatchSnapshot();
      expect((vfile.data as any).markdownDI?.errors).toBeDefined();
    });

    test('handles invalid yaml in frontmatter', async () => {
      const input = `---
invalid yaml: : : here
---

# Content
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(String(vfile)).toMatchSnapshot();
    });

    test('handles file read errors gracefully', async () => {
      const input = `---
name: file-error-test
description: Testing file read errors

blueprints:
  sections:
    missing: sections/does-not-exist.md
---

{{sections.missing}}
`;

      const processor = unified()
        .use(remarkParse)
        .use(remarkFrontmatter, ['yaml'])
        .use(remarkMarkdownDI, { baseDir: TEST_DIR })
        .use(remarkStringify);

      const vfile = await processor.process(input);

      expect(String(vfile)).toMatchSnapshot();
      expect((vfile.data as any).markdownDI?.errors.some((e: any) => e.type === 'file')).toBe(true);
    });
  });
});
