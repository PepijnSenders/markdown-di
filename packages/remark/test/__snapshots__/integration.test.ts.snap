// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`remarkMarkdownDI - Integration Tests Basic Remark Processing processes markdown with dependency injection via remark 1`] = `
"---
name: remark-test
description: Testing remark plugin

blueprints:
  sections:
    header: sections/header.md
---

{{sections.header}}

## Main Content

This is the main content.
"
`;

exports[`remarkMarkdownDI - Integration Tests Basic Remark Processing processes document with multiple transclusions 1`] = `
"---
name: multi-transclusion
description: Multiple transclusions test

blueprints:
  sections:
    header: sections/header.md
    footer: sections/footer.md
  components:
    callout: components/callout.md
---

{{sections.header}}

## Content Section

Here's some content with a callout:

{{components.callout}}

More content here.

{{sections.footer}}
"
`;

exports[`remarkMarkdownDI - Integration Tests Basic Remark Processing handles code blocks in transclusions 1`] = `
"---
name: code-transclusion
description: Testing code block transclusion

blueprints:
  components:
    example: components/code-example.md
---

# Code Examples

Here's a code example:

{{components.example}}

That's the example.
"
`;

exports[`remarkMarkdownDI - Integration Tests Validation Mode validates without processing in validate mode 1`] = `
"---
name: validation-test
description: Testing validation mode

blueprints:
  sections:
    header: sections/header.md
    nonexistent: sections/does-not-exist.md
---

{{sections.header}}
{{sections.nonexistent}}
"
`;

exports[`remarkMarkdownDI - Integration Tests Validation Mode validates without processing in validate mode 2`] = `
{
  "markdownDI": {
    "dependencies": [
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/sections/header.md",
    ],
    "errors": [
      {
        "location": "blueprints.sections.nonexistent",
        "message": "Blueprint file not found: sections/does-not-exist.md",
        "type": "file",
      },
    ],
    "frontmatter": {
      "blueprints": {
        "sections": {
          "header": "sections/header.md",
          "nonexistent": "sections/does-not-exist.md",
        },
      },
      "description": "Testing validation mode",
      "name": "validation-test",
    },
  },
}
`;

exports[`remarkMarkdownDI - Integration Tests Validation Mode collects validation errors in file data 1`] = `
{
  "markdownDI": {
    "dependencies": [],
    "errors": [
      {
        "location": "line 2, column 0",
        "message": "Reference "{{undefined.reference}}" is not declared in frontmatter",
        "type": "reference",
      },
      {
        "location": "line 3, column 0",
        "message": "Reference "{{another.missing}}" is not declared in frontmatter",
        "type": "reference",
      },
    ],
    "frontmatter": {
      "description": "Collecting validation errors",
      "name": "error-collection",
    },
  },
}
`;

exports[`remarkMarkdownDI - Integration Tests Build Mode processes and replaces references in build mode 1`] = `
"---
name: build-mode-test
description: Testing build mode processing

blueprints:
  sections:
    header: sections/header.md
    footer: sections/footer.md
---

{{sections.header}}

# Main Section

Content goes here.

{{sections.footer}}
"
`;

exports[`remarkMarkdownDI - Integration Tests Advanced Options respects headingShift option 1`] = `
"---
name: heading-shift-test
description: Testing heading shift

blueprints:
  sections:
    header: sections/header.md
---

# Top Level

{{sections.header}}
"
`;

exports[`remarkMarkdownDI - Integration Tests Advanced Options respects linkRewrite option 1`] = `
"---
name: link-rewrite-test
description: Testing link rewriting

blueprints:
  sections:
    header: sections/header.md
---

# Document

{{sections.header}}
"
`;

exports[`remarkMarkdownDI - Integration Tests File Metadata stores frontmatter in file data 1`] = `
{
  "markdownDI": {
    "dependencies": [
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/sections/header.md",
    ],
    "errors": [],
    "frontmatter": {
      "author": "Test Author",
      "blueprints": {
        "sections": {
          "header": "sections/header.md",
        },
      },
      "description": "Testing metadata storage",
      "name": "metadata-test",
    },
  },
}
`;

exports[`remarkMarkdownDI - Integration Tests File Metadata stores dependencies in file data 1`] = `
{
  "markdownDI": {
    "dependencies": [
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/sections/header.md",
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/sections/footer.md",
    ],
    "errors": [],
    "frontmatter": {
      "blueprints": {
        "sections": {
          "footer": "sections/footer.md",
          "header": "sections/header.md",
        },
      },
      "description": "Testing dependency tracking",
      "name": "dependencies-test",
    },
  },
}
`;

exports[`remarkMarkdownDI - Integration Tests File Metadata stores errors in file data 1`] = `
{
  "markdownDI": {
    "dependencies": [
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/sections/header.md",
    ],
    "errors": [
      {
        "location": "line 3, column 0",
        "message": "Reference "{{sections.undefined}}" is not declared in frontmatter",
        "type": "reference",
      },
    ],
    "frontmatter": {
      "blueprints": {
        "sections": {
          "header": "sections/header.md",
        },
      },
      "description": "Testing error tracking",
      "name": "errors-test",
    },
  },
}
`;

exports[`remarkMarkdownDI - Integration Tests Complex Scenarios processes complex document with nested structures 1`] = `
"---
name: complex-document
description: Complex document with various elements

blueprints:
  sections:
    header: sections/header.md
    footer: sections/footer.md
  components:
    callout: components/callout.md
    code: components/code-example.md
---

{{sections.header}}

## Introduction

This is a complex document demonstrating various features.

{{components.callout}}

## Code Examples

{{components.code}}

## Conclusion

Thank you for reading.

{{sections.footer}}
"
`;

exports[`remarkMarkdownDI - Integration Tests Complex Scenarios processes complex document with nested structures 2`] = `
{
  "markdownDI": {
    "dependencies": [
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/sections/header.md",
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/sections/footer.md",
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/components/callout.md",
      "/Users/ps/Documents/GitHub/markdown-di/packages/test-fixtures-remark/components/code-example.md",
    ],
    "errors": [],
    "frontmatter": {
      "blueprints": {
        "components": {
          "callout": "components/callout.md",
          "code": "components/code-example.md",
        },
        "sections": {
          "footer": "sections/footer.md",
          "header": "sections/header.md",
        },
      },
      "description": "Complex document with various elements",
      "name": "complex-document",
    },
  },
}
`;

exports[`remarkMarkdownDI - Integration Tests Complex Scenarios handles empty documents gracefully 1`] = `
"---
name: empty-document
description: Empty document test
---
"
`;

exports[`remarkMarkdownDI - Integration Tests Complex Scenarios preserves markdown formatting 1`] = `
"---
name: formatting-preservation
description: Testing formatting preservation

blueprints:
  components:
    callout: components/callout.md
---

# Main Title

This is **bold** and this is *italic*.

{{components.callout}}

* List item 1
* List item 2
  * Nested item
* List item 3

1. Numbered list
2. Second item
   1. Nested numbered
3. Third item
"
`;

exports[`remarkMarkdownDI - Integration Tests Error Handling handles missing frontmatter 1`] = `
"# Document Without Frontmatter

{{sections.header}}
"
`;

exports[`remarkMarkdownDI - Integration Tests Error Handling handles invalid yaml in frontmatter 1`] = `
"---
invalid yaml: : : here
---

# Content
"
`;

exports[`remarkMarkdownDI - Integration Tests Error Handling handles file read errors gracefully 1`] = `
"---
name: file-error-test
description: Testing file read errors

blueprints:
  sections:
    missing: sections/does-not-exist.md
---

{{sections.missing}}
"
`;
